pkg/process/monitor/process_monitor.go

```diff
@@ -0,0 +1,285 @@
+
+	initializing bool
+
+	events chan netlink.ProcEvent
+	done   chan struct{}
+	errors chan error
+
+	procEventCallbacks map[ProcessEventType][]*ProcessCallback
+	callbackRunner     chan func()
+	callbackRunnerDone chan struct{}
+}
+
+type ProcessEventType int
+
+const (
+	EXEC ProcessEventType = iota
+	FORK
+	EXIT
+)
+
+type ProcessMetadataField int
+
+const (
+	ANY ProcessMetadataField = iota
+	NAME
+	MAPFILE
+)
+
+type ProcessCallback struct {
+	Event    ProcessEventType
+	Metadata ProcessMetadataField
+	Regex    *regexp.Regexp
+	Callback func(pid uint32)
+}
+
+func (pm *ProcessMonitor) enqueueCallback(callback *ProcessCallback, pid uint32) {
+	pm.callbackRunner <- func() { callback.Callback(pid) }
+}
+
+// Initialize() will scan all running processes and execute matching callbacks
+// Once it's done all new events from netlink socket will be processed by the main async loop
+func (pm *ProcessMonitor) Initialize() error {
+	fn := func(pid int) error {
+		for _, c := range pm.procEventCallbacks[EXEC] {
+			pm.evalEXECCallback(c, uint32(pid))
+		}
+		return nil
+	}
+
+	pm.m.Lock()
+	defer pm.m.Unlock()
+	if err := util.WithAllProcs(util.HostProc(), fn); err != nil {
+		return fmt.Errorf("process monitor init, scanning all process failed %s", err)
+	}
+	// enable events to be processed
+	pm.initializing = false
+	return nil
+}
+
+func (p *ProcessMonitor) evalEXECCallback(c *ProcessCallback, pid uint32) {
+	if c.Metadata == ANY {
+		p.enqueueCallback(c, pid)
+		return
+	}
+
+	proc, err := process.NewProcess(int32(pid))
+	if err != nil {
+		log.Errorf("process %d parsing failed %s", pid, err)
+		return
+	}
+	switch c.Metadata {
+	case NAME:
+		pname, err := proc.Name()
+		if err != nil {
+			log.Errorf("process %d name parsing failed %s", pid, err)
+			return
+		}
+		if c.Regex.MatchString(pname) {
+			p.enqueueCallback(c, pid)
+		}
+	case MAPFILE:
+		mmaps, err := proc.MemoryMaps(true)
+		if err != nil {
+			log.Errorf("process %d maps parsing failed %s", pid, err)
+			return
+		}
+		if mmaps == nil {
+			return
+		}
+		for _, mmap := range *mmaps {
+			if c.Regex.MatchString(mmap.Path) {
+				p.enqueueCallback(c, pid)
+				break
+			}
+		}
+	}
+}
+
+func (pm *ProcessMonitor) Subscribe(callback *ProcessCallback) (UnSubscribe func(), err error) {
+	pm.m.Lock()
+	defer pm.m.Unlock()
+
+	for _, c := range pm.procEventCallbacks[callback.Event] {
+		if c == callback {
+			return func() {}, errors.New("same callback can't be registred twice")
+		}
+	}
+
+	pm.procEventCallbacks[callback.Event] = append(pm.procEventCallbacks[callback.Event], callback)
+
+	// UnSubscribe()
+	return func() {
+		pm.m.Lock()
+		defer pm.m.Unlock()
+
+		for i, c := range pm.procEventCallbacks[callback.Event] {
+			if c == callback {
+				l := len(pm.procEventCallbacks[callback.Event])
+				pm.procEventCallbacks[callback.Event][i] = pm.procEventCallbacks[callback.Event][l-1]
+				pm.procEventCallbacks[callback.Event] = pm.procEventCallbacks[callback.Event][:l-1]
```

Comment:
can you add a comment to explain what are you doing inside the for loop? or even better add a short comment for the entire for loop to explain how does Subsribe process work